bigtop-packages/src/charm/layer-hive/lib/charms/layer/apache_hive.py

```diff
@@ -0,0 +1,137 @@
+
+    def install(self):
+        '''
+        Trigger the Bigtop puppet recipe that handles the Hive service.
+        '''
+        # Dirs are handled by the bigtop deb. No need to call out to
+        # dist_config to do that work.
+        self.dist_config.add_users()
+        roles = ['hive-client']
+
+        bigtop = Bigtop()
+        bigtop.render_site_yaml(roles=roles)
+        bigtop.trigger_puppet()
+
+    def initial_hive_config(self):
+        '''
+        Configure system-wide hive bits and get hive config files set up
+        so future config changes can be added where needed.
+        '''
+        # Configure system-wide bits
+        hive_bin = self.dist_config.path('hive') / 'bin'
+        with utils.environment_edit_in_place('/etc/environment') as env:
+            if hive_bin not in env['PATH']:
+                env['PATH'] = ':'.join([env['PATH'], hive_bin])
+            env['HIVE_CONF_DIR'] = self.dist_config.path('hive_conf')
+
+        # Copy templates to config files if they don't exist
+        hive_env = self.dist_config.path('hive_conf') / 'hive-env.sh'
+        hive_log4j = self.dist_config.path('hive_conf') / 'hive-log4j.properties'
+        hive_site = self.dist_config.path('hive_conf') / 'hive-site.xml'
+        if not hive_env.exists():
+            (self.dist_config.path('hive_conf') / 'hive-env.sh.template').copy(hive_env)
+        if not hive_site.exists():
+            (self.dist_config.path('hive_conf') / 'hive-default.xml.template').copy(hive_site)
+
+        # Configure immutable things
+        hive_logs = self.dist_config.path('hive_logs')
+        utils.re_edit_in_place(hive_log4j, {
+            r'^hive.log.dir.*': 'hive.log.dir={}'.format(hive_logs),
+        })
+        with utils.xmlpropmap_edit_in_place(hive_site) as props:
+            # XXX (kwm): these 4 were needed in 0.12, but it seems ok without them in > 1.0
+            # props['hive.exec.local.scratchdir'] = "/tmp/hive"
+            # props['hive.downloaded.resources.dir'] = "/tmp/hive_resources"
+            # props['hive.querylog.location'] = "/tmp/hive"
+            # props['hive.server2.logging.operation.log.location'] = "/tmp/hive"
+            props['hive.hwi.war.file'] = "lib/hive-hwi.jar"
+
+    # called during config-changed events
+    def configure_hive(self):
+        config = hookenv.config()
+        hive_env = self.dist_config.path('hive_conf') / 'hive-env.sh'
+        utils.re_edit_in_place(hive_env, {
+            r'.*export HADOOP_HEAPSIZE *=.*': 'export HADOOP_HEAPSIZE=%s' % config['heap'],
+        })
+
+    def configure_remote_db(self, mysql):
+        hive_site = self.dist_config.path('hive_conf') / 'hive-site.xml'
+        jdbc_url = "jdbc:mysql://{}:{}/{}?createDatabaseIfNotExist=true".format(
+            mysql.host(), mysql.port(), mysql.database()
+        )
+        with utils.xmlpropmap_edit_in_place(hive_site) as props:
```

Comment:
Lets do this configuration inside bigtop. It would be very beneficial for bigtop to use a DB other than derby